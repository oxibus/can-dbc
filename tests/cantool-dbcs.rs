#![cfg(feature = "with-serde")]

use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use std::fs::File;
use std::io::Read as _;
use std::path::Path;
use std::{fs, io};

/// Test parsing all DBC files in the `tests/cantools-dbcs` directory, using the `can_dbc` crate.
#[test]
fn test_cantools_dbc_files() {
    // Configure insta snapshot content and location
    with_settings! {
        { omit_expression => true,
          snapshot_path => "snapshots-cantools-dbcs",
          prepend_module_to_snapshot => false },
        {
            test_cantools_dbc_files_int().unwrap();
        }
    }
}

/// Test parsing all DBC files in the `tests/cantools-dbcs` directory.
fn test_cantools_dbc_files_int() -> io::Result<()> {
    for dbc_path in fs::read_dir("./tests/cantools-dbcs")? {
        let path = dbc_path?.path();
        if path
            .extension()
            .and_then(|s| s.to_str())
            .map(|s| s.eq_ignore_ascii_case("dbc"))
            .is_some()
        {
            parse_one_file(&path)?;
        }
    }
    Ok(())
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: &Path) -> io::Result<()> {
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    eprintln!("Testing DBC file: {file_name}.dbc");
    let mut buffer = Vec::new();
    File::open(path)?.read_to_end(&mut buffer)?;

    match can_dbc::DBC::from_slice(&buffer) {
        Ok(dbc) => assert_yaml_snapshot!(file_name, dbc),
        Err(e) => {
            eprintln!("Failed to parse {file_name}.dbc: {e:?}");
            assert_debug_snapshot!(format!("!error___{file_name}"), e);
        }
    }
    Ok(())
}
