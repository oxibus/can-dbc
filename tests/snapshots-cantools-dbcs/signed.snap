---
source: tests/cantool-dbcs.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - []
value_tables: []
messages:
  - message_id:
      Standard: 10
    message_name: Message378910
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s3big
        multiplexer_indicator: Plain
        start_bit: 39
        signal_size: 3
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: s3
        multiplexer_indicator: Plain
        start_bit: 34
        signal_size: 3
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: s10big
        multiplexer_indicator: Plain
        start_bit: 40
        signal_size: 10
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: s8big
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 8
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: s7big
        multiplexer_indicator: Plain
        start_bit: 62
        signal_size: 7
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: s9
        multiplexer_indicator: Plain
        start_bit: 17
        signal_size: 9
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: s8
        multiplexer_indicator: Plain
        start_bit: 26
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: s7
        multiplexer_indicator: Plain
        start_bit: 1
        signal_size: 7
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 9
    message_name: Message63big_1
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s63big
        multiplexer_indicator: Plain
        start_bit: 6
        signal_size: 63
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 8
    message_name: Message63_1
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s63
        multiplexer_indicator: Plain
        start_bit: 1
        signal_size: 63
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 7
    message_name: Message63big
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s63big
        multiplexer_indicator: Plain
        start_bit: 7
        signal_size: 63
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 6
    message_name: Message63
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s63
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 63
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 5
    message_name: Message32big
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s32big
        multiplexer_indicator: Plain
        start_bit: 7
        signal_size: 32
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 4
    message_name: Message33big
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s33big
        multiplexer_indicator: Plain
        start_bit: 7
        signal_size: 33
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 3
    message_name: Message64big
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s64big
        multiplexer_indicator: Plain
        start_bit: 7
        signal_size: 64
        byte_order: BigEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 2
    message_name: Message64
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s64
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 64
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: -9223372036854780000
        max: 9223372036854780000
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 1
    message_name: Message33
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s33
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 33
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: -4294967296
        max: 4294967295
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 0
    message_name: Message32
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: s32
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 32
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
message_transmitters: []
environment_variables: []
environment_variable_data: []
signal_types: []
comments: []
attribute_definitions:
  - Plain: "\"BusType\" STRING "
attribute_defaults:
  - attribute_name: BusType
    attribute_value:
      AttributeValueCharString: ""
attribute_values: []
value_descriptions: []
signal_type_refs: []
signal_groups: []
signal_extended_value_type_list: []
extended_multiplex: []
